streams/src/main/java/org/apache/kafka/streams/state/internals/BoundsCheckingWindowStore.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.streams.kstream.Windowed;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.internals.InternalProcessorContext;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.WindowStore;
+import org.apache.kafka.streams.state.WindowStoreIterator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BoundsCheckingWindowStore<K, V> implements WindowStore<K, V> {
+    private static final Logger LOG = LoggerFactory.getLogger(BoundsCheckingWindowStore.class);
+
+    private final WindowStore<K, V> delegate;
+    private final long retention;
+
+    private ProcessorContext context;
+
+    public BoundsCheckingWindowStore(final WindowStore<K, V> delegate, final long retention) {
+        this.delegate = delegate;
+        this.retention = retention;
+    }
+
+    @Override
+    public void put(final K key, final V value) {
+        delegate.put(key, value);
```

Comment:
For me, I'm not sure as does this imply a change in our widowing semantics?  Having said that, I'm not opposed to the change either.

Reply:
Yeah, I wasn't sure about it either. Its only use is in KStreamJoinWindow, where its function is to window the incoming streams for joining.

I'm not sure, but I think this would be a semantic improvement. One that probably is out of scope for this ticket, though...

Reply:
I think we should use the record timestamp than the stream time here for stream-stream join window (note that the usage of the window store is not the best choice here for stream-stream join, but we're where we're for now). Note that when we put a record into the corresponding window store, it should be joinable with any record in the other stream with timestamp in [t - join-window, t + join-window], where `t` is the record timestamp here. I.e. we are a bit mis-using the window store in this way, but using record timestamp is the better choice than stream time in this case.

Reply:
The JavaDoc does not seem to be correct. `RocksDBWindowStore` uses `context.timestamp()` -- also, this timestamp is encoded in the key. I think (but not 100% sure) we need this for stream-stream joins as this timestamp tells us if a record is join-able or too far apart. Thus, changing to stream-time would be incorrect.

However, I am wondering atm: if we use the same store for window-aggregation and for joins, than we do store the timestamp of the record in the store already. I think I am missing something here, as I thought that we only store the plain key, or windowed-key for aggregations... \cc @guozhangwang @dguy

Reply:
Ok, I'll update the JavaDoc to specify that it'll use record time.

Reply:
Confirmed, all implementations of this method use record time.

Your question about joins is applicable. The only usage of this method in production code is in KStreamJoinWindow.java.

Reply:
`than we do store the timestamp of the record in the store already` not sure if I understand fully about this, could you elaborate?